app/Http/Controllers/Auth/AuthController.php

```diff
@@ -267,14 +183,145 @@ public function emailChangeStep3($token)
+            Auth::user()->generateApiKey();
+
+            return self::json(Auth::user()->load('AdserverWallet')->toArray(), 200);
+        }
+
+        return self::json([], 400);
+    }
+
+    /**
+     * Log the user out of the application.
+     *
+     * @param \Illuminate\Http\Request $request
+     *
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function logout(Request $request)
+    {
+        Auth::user()->clearApiKey();
+
+        return self::json([], 204);
+    }
+
+    /**
+     * Start password recovery process - generate and send email.
+     *
+     * @param \Illuminate\Http\Request $request
+     *
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function recovery(Request $request)
+    {
+        Validator::make($request->all(), ['email' => 'required|email', 'uri' => 'required'])->validate();
+        $user = User::where('email', $request->input('email'))->first();
+        if (empty($user)) {
+            return self::json([], 204);
+        }
+        DB::beginTransaction();
+        if (!Token::canGenerate($user->id, 'password-recovery', $this->password_recovery_resend_limit)) {
+            return self::json([], 204);
+        }
+        Mail::to($user)->queue(new AuthRecovery(
+            Token::generate('password-recovery', $this->password_recovery_token_time, $user->id),
+            $request->input('uri')
+        ));
+        DB::commit();
+
+        return self::json([], 204);
+    }
+
+    /**
+     * Tests and extends user password recovery token.
+     *
+     * @return \Illuminate\Http\JsonResponse
+     */
+    public function recoveryTokenExtend($token)
+    {
+        if (Token::extend($token, $this->password_recovery_token_time, null, 'password-recovery')) {
+            return self::json([], 204);
+        }
+
+        return self::json([], 422, ['message' => 'Password recovery token is invalid']);
     }
 
-    public function read(Request $request, $user_id)
+    public function updateSelf(Request $request)
     {
-        // TODO check privileges
-        $user = User::findOrFail($user_id);
+        if (!Auth::check() && !$request->has('user.token')) {
+            return self::json([], 401, ['message' => 'Required authenticated access or token authentication']);
+        }
+
+        DB::beginTransaction();
+        if (Auth::check()) {
+            $user = Auth::user();
+            $token_authorization = false;
+        } else {
+            if (false === $token = Token::check($request->input('user.token'), null, 'password-recovery')) {
+                DB::rollBack();
+
+                return self::json([], 422, ['message' => 'Authentication token is invalid']);
+            }
+            $user = User::findOrFail($token['user_id']);
+            $token_authorization = true;
+        }
+
+        $this->validateRequestObject($request, 'user', User::$rules);
+        $user->fill($request->input('user'));
+
+        if (!$request->has('user.password_new')) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "user.password_new" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1192)